sys/net/application_layer/gcoap/gcoap.c

```diff
@@ -732,6 +881,95 @@ static void _find_obs_memo_resource(gcoap_observe_memo_t **memo,
     }
 }
 
+/*
+ * Transport layer functions
+ */
+
+static void _tl_get_socket(coap_socket_t *sock)
+{
+#if IS_ACTIVE(CONFIG_GCOAP_ENABLE_DTLS)
+    sock->type = COAP_SOCKET_TYPE_DTLS;
+    sock->socket.dtls = &_sock_dtls;
+#else
+    sock->type = COAP_SOCKET_TYPE_UDP;
+    sock->socket.udp = &_sock_udp;
+#endif
+}
+
+static ssize_t _tl_send(coap_socket_t *sock, const void *data, size_t len,
+                      const sock_udp_ep_t *remote)
+{
+    int res;
+#if IS_ACTIVE(CONFIG_GCOAP_ENABLE_DTLS)
+    /* prepare session */
+    sock_dtls_session_set_udp_ep(&sock->ctx_dtls_session, remote);
+    dsm_state_t session_state = dsm_store(sock->socket.dtls, &sock->ctx_dtls_session,
+                  SESSION_STATE_HANDSHAKE, true);
+    if (session_state == NO_SPACE) {
+        return -1;
+    }
+
+    /* send application data */
+    res = sock_dtls_send(sock->socket.dtls, &sock->ctx_dtls_session, data, len,
+                            SOCK_NO_TIMEOUT);
+    if (res <= 0 ) {
+        dsm_remove(sock->socket.dtls, &sock->ctx_dtls_session);
+        sock_dtls_session_destroy(sock->socket.dtls, &sock->ctx_dtls_session);
+    }
+#else
+    res = sock_udp_send(sock->socket.udp, data, len, remote);
+#endif
+    return res;
+}
+
+#if IS_ACTIVE(CONFIG_GCOAP_ENABLE_DTLS)
+static ssize_t _tl_authenticate(coap_socket_t *sock, const sock_udp_ep_t *remote,
+                                uint32_t timeout)
+{
+    int res;
+
+    /* prepare session */
+    sock_dtls_session_set_udp_ep(&sock->ctx_dtls_session, remote);
+    dsm_state_t session_state = dsm_store(sock->socket.dtls, &sock->ctx_dtls_session,
+                                          SESSION_STATE_HANDSHAKE, true);
+    if (session_state == SESSION_STATE_ESTABLISHED) {
+        return 0;
+    }
+    if (session_state == NO_SPACE) {
+        return -1;
+    }
+
+    /* start handshake */
+    _auth_waiting_thread = thread_getpid();
+    res = sock_dtls_session_init(sock->socket.dtls, remote, &sock->ctx_dtls_session);
+    if (res == 0) {
+        _auth_waiting_thread = -1;
+        return res;
+    }
+
+    msg_t msg;
+    bool is_timed_out = false;
+    do {
+        uint32_t start = xtimer_now_usec();
+        res = xtimer_msg_receive_timeout(&msg, timeout);
+
+        if (timeout != SOCK_NO_TIMEOUT) {
+            uint32_t diff = (xtimer_now_usec() - start);
+            timeout = (diff > timeout) ? 0: timeout - diff;
+            is_timed_out = (res < 0) || (timeout == 0);
+        }
```

Comment:
What is going on here?

Reply:
This is the mechanism from sock_dtls for blocking handshake in `sock_dtls_send()`. For the case that we receive another msg than the expected in this time, we wait the full timeout time to receive `DTLS_CONNECT`.

Reply:
Can you add a comment about this?
Future readers won't be hunting down the PR comment :wink: